From cce67db61663d0908cdc181082f9bba274359bdb Mon Sep 17 00:00:00 2001
From: Kamil Renczewski <kamil.renczewski@codilime.com>
Date: Thu, 5 Oct 2017 14:17:54 +0200
Subject: [PATCH 08/15] WIP: Handle all values of `required` field

---
 terraform_mappings.py | 15 +++++++++++----
 1 file changed, 11 insertions(+), 4 deletions(-)

diff --git a/terraform_mappings.py b/terraform_mappings.py
index f3426c7..6dbdcd1 100644
--- a/terraform_mappings.py
+++ b/terraform_mappings.py
@@ -178,16 +178,23 @@ func %(camel)s() map[string]*schema.Schema {
         file.write(decl)
 
         for member in ctype.getDataMembers():
-            #print "Member dump: " + dump(member.xsd_object) + "\m"
             elem = ""
             req = ''
             mtype = TerraformTypeName(member.xsd_object.getType(),
                                           orig_name = True)
-            if member.xsd_object.required == True:
+            if member.xsd_object.required is None:
+                req = 'false'
+            elif str(member.xsd_object.required) == 'required':
+                req = 'true'
+            elif str(member.xsd_object.required) == 'true':
                 req = 'true'
-            elif member.xsd_object.required == False:
+            elif str(member.xsd_object.required) == 'false':
+                req = 'false'
+            elif str(member.xsd_object.required) == 'optional':
                 req = 'false'
-            elif member.xsd_object.required == 'optional':
+            elif str(member.xsd_object.required) == 'exclusive':
+                req = 'false' # Only one of values need to exist
+            elif str(member.xsd_object.required) == 'system-only':
                 req = 'false'
             else:
                 print "For " + ctype.getName() + "->" + member.membername +\
-- 
2.7.4

